### PR TITLE
Fixed App error on pdf to image for protected files

### LINKED ISSUE #727: The app crashed when I tried to convert a password protected PDF file to Image
#### Expected behavior
The users should be able to convert a password protected PDF file to Image without crashing the app.

#### Actual behavior
When I tried to convert a password protected PDF file to Image, sometimes it crashed immediately/ after a few attempts.

#### How to reproduce
1. Open the app.
2. Then go to "PDF to Images".
3. Swipe up the view files menu.
4. Choose and click on a password protected PDF file. (If you don't have any password protected file then download from somewhere or create a new one from "Add password" menu).
5. Then click on "CREATE IMAGES" and notice the crash. (Sometimes it crashes after a few attempts.)

* App version: 8.6.5
* Operating system: Android 6.0.1

#### Recording Of The Bug
imagetopdf

#### LogCat
```
04-22 12:15:35.743 22730 22788 E AndroidRuntime: FATAL EXCEPTION: AsyncTask #4

04-22 12:15:35.743 22730 22788 E AndroidRuntime: Process: swati4star.createpdf, PID: 22730

04-22 12:15:35.743 22730 22788 E AndroidRuntime: java.lang.RuntimeException: An error occurred while executing doInBackground()

04-22 12:15:35.743 22730 22788 E AndroidRuntime: 	at android.os.AsyncTask$3.done(AsyncTask.java:309)

04-22 12:15:35.743 22730 22788 E AndroidRuntime: 	at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:354)

04-22 12:15:35.743 22730 22788 E AndroidRuntime: 	at java.util.concurrent.FutureTask.setException(FutureTask.java:223)

04-22 12:15:35.743 22730 22788 E AndroidRuntime: 	at java.util.concurrent.FutureTask.run(FutureTask.java:242)

04-22 12:15:35.743 22730 22788 E AndroidRuntime: 	at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:234)

04-22 12:15:35.743 22730 22788 E AndroidRuntime: 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1113)

04-22 12:15:35.743 22730 22788 E AndroidRuntime: 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:588)

04-22 12:15:35.743 22730 22788 E AndroidRuntime: 	at java.lang.Thread.run(Thread.java:818)

04-22 12:15:35.743 22730 22788 E AndroidRuntime: Caused by: java.lang.IllegalStateException: cannot load page

04-22 12:15:35.743 22730 22788 E AndroidRuntime: 	at android.graphics.pdf.PdfRenderer.nativeOpenPageAndGetSize(Native Method)

04-22 12:15:35.743 22730 22788 E AndroidRuntime: 	at android.graphics.pdf.PdfRenderer.access$300(PdfRenderer.java:101)

04-22 12:15:35.743 22730 22788 E AndroidRuntime: 	at android.graphics.pdf.PdfRenderer$Page.<init>(PdfRenderer.java:276)

04-22 12:15:35.743 22730 22788 E AndroidRuntime: 	at android.graphics.pdf.PdfRenderer$Page.<init>(PdfRenderer.java:254)

04-22 12:15:35.743 22730 22788 E AndroidRuntime: 	at android.graphics.pdf.PdfRenderer.openPage(PdfRenderer.java:203)

04-22 12:15:35.743 22730 22788 E AndroidRuntime: 	at swati4star.createpdf.util.PdfToImages.doInBackground(PdfToImages.java:62)

04-22 12:15:35.743 22730 22788 E AndroidRuntime: 	at swati4star.createpdf.util.PdfToImages.doInBackground(PdfToImages.java:22)

04-22 12:15:35.743 22730 22788 E AndroidRuntime: 	at android.os.AsyncTask$2.call(AsyncTask.java:295)

04-22 12:15:35.743 22730 22788 E AndroidRuntime: 	at java.util.concurrent.FutureTask.run(FutureTask.java:237)

04-22 12:15:35.743 22730 22788 E AndroidRuntime: 	... 4 more
```

### PR DESCRIPTION
**Pull request by jddeep**
# Description
Fixed App error on pdf to image for protected files.

Fixes #727 

## Type of change
Just put an x in the [] which are valid.
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes.
- [x] `./gradlew assembleDebug assembleRelease`
- [x] `./gradlew checkstyle`

# Checklist:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings

### CODE REVIEW

**Comment by jddeep** [comment]
@vyankatesh24 Please review.

**Comment by sidhuparas** [comment]
@jddeep I couldn't replicate the issue despite trying many times on a password-protected PDF file. I am getting `SecurityException: incorrect password...` in logcat but the app isn't crashing and snackbar shows that no image files were found.

**Comment by jddeep** [comment]
> @jddeep I couldn't replicate the issue despite trying many times on a password-protected PDF file. I am getting `SecurityException: incorrect password...` in logcat but the app isn't crashing and snackbar shows that no image files were found.

Yes @sidhuparas The app won't crash. Here, there can be two reasons why you are facing this error. Either your selected pdf has no images in it or the password you entered to decrypt the file is wrong(likely this is the problem).

File: app/src/main/java/swati4star/createpdf/util/ExtractImages.java [hunk 1]
```diff
@@ -23,11 +23,28 @@
     private ExtractImagesListener mExtractImagesListener;
     private int mImagesCount = 0;
     private ArrayList<String> mOutputFilePaths;
+    private PDFEncryptionUtility mPdfEncryptionUtility;
+    private PDFUtils mPDFUtils;
 
-    public ExtractImages(String mPath, ExtractImagesListener mExtractImagesListener) {
+    public ExtractImages(String mPath, PDFEncryptionUtility pdfEncryptionUtility,
+                         PDFUtils pdfUtils, ExtractImagesListener mExtractImagesListener) {
```
Comment (Swati4star) [inline app/src/main/java/swati4star/createpdf/util/ExtractImages.java]: we shouldnt take `pdfUtils` or `pdfEncryptionUtility` as parameters to this function. we can create instance of these classes here only. Just take "Context" as the paramter

File: app/src/main/java/swati4star/createpdf/util/PdfToImages.java [hunk 2]
```diff
@@ -26,12 +26,29 @@
     private ExtractImagesListener mExtractImagesListener;
     private int mImagesCount = 0;
     private ArrayList<String> mOutputFilePaths;
+    private PDFEncryptionUtility mPdfEncryptionUtility;
+    private PDFUtils mPDFUtils;
 
-    public PdfToImages(String mPath, Uri mUri, ExtractImagesListener mExtractImagesListener) {
+    public PdfToImages(String mPath, Uri mUri, PDFEncryptionUtility pdfEncryptionUtility,
```
Comment (Swati4star) [inline app/src/main/java/swati4star/createpdf/util/PdfToImages.java]: same thing here

**Comment by jddeep** [comment]
@Swati4star Done!

**Comment by jddeep** [comment]
@vyankatesh24 Some pdfs i am able to convert to images while some are showing similar problem. Can you please test into it and suggest what could be the problem now.

**Comment by sidhuparas** [comment]
@jddeep It didn't ask me for password to decrypt. But if it isn't able to decrypt PDF, how can it scan for images?

**Comment by jddeep** [comment]
@sidhuparas It would ask for password. Please check again. Pull the latest commit and try again.

**Comment by sidhuparas** [comment]
@jddeep Actually I didn't even try this PR's code. Master branch code is working fine for me without any crash. Tried for several different files and tried many times.

**Comment by jddeep** [comment]
@sidhuparas i had the same confusion but @vyankatesh24 pointed it out. Try to extract images from password protected files. It would show error snackbar.

File: app/src/main/java/swati4star/createpdf/util/ExtractImages.java [hunk 3]
```diff
@@ -23,25 +25,51 @@
...
+    }
+
+    /**
+     * This function checks whether a file is encrypted or not and if it is
+     * then it decrypts it.
+     *
+     * @param path is the path of the file which needs to be checked for encryption.
+     */
+    private void decryptFile(String path) {
+        if (mPDFUtils.isPDFEncrypted(path)) {
```
Comment (jddeep) [inline app/src/main/java/swati4star/createpdf/util/ExtractImages.java]: @vyankatesh24 The pdf should be decrypted that is dialog should appear only if it is encrypted. Why is that for every pdf?

Comment (vyankatesh24) [reply app/src/main/java/swati4star/createpdf/util/ExtractImages.java]: Dialog should not appear for every PDF.  Only for encrypted ones.

Comment (jddeep) [reply app/src/main/java/swati4star/createpdf/util/ExtractImages.java]: Yes that is what I am saying.

Comment (jddeep) [reply app/src/main/java/swati4star/createpdf/util/ExtractImages.java]: The logic is according to that only right? It should enter the statements in the if-block if the file is Encrypted.

Comment (jddeep) [reply app/src/main/java/swati4star/createpdf/util/ExtractImages.java]: @vyankatesh24 ?

**Comment by jddeep** [comment]
@vyankatesh24 I have tested it. The password dialog only appears for encrypted files and not for all files. Have you tested my last and latest changes?

File: app/src/main/java/swati4star/createpdf/util/PdfToImages.java [hunk 4]
```diff
@@ -48,13 +74,18 @@ protected Void doInBackground(Void... voids) {
...
         try {
-            if (mUri != null)
-                // resolve pdf file path based on uri
-                fileDescriptor = ((PdfToImageFragment) mExtractImagesListener).getContext()
-                        .getContentResolver().openFileDescriptor(mUri, "r");
-            else if (mPath != null)
-                // resolve pdf file path based on relative path
-                fileDescriptor = ParcelFileDescriptor.open(new File(mPath), MODE_READ_ONLY);
+            if (mDecryptedPath != null)
+                fileDescriptor = ParcelFileDescriptor.open(new File(mDecryptedPath), MODE_READ_ONLY);
```
Comment (jddeep) [inline app/src/main/java/swati4star/createpdf/util/PdfToImages.java]: And @vyankatesh24 The only problem i am facing is that the `mDecryptedPath` is coming `null` the first time. I mean how sync this after `mDecryptedPath = mPdfEncryptionUtility.removePasswordForImages(path, null);` gets executed? The file is indeed getting decrypted and a new decrypted path with a new path is getting created but as the processes are not in sync, It seems `doInBackground` works ahead. But It should execute after `onPreExecute` right? @vyankatesh24 Please check through and let me know.

**Comment by vyankatesh24** [comment]
> @vyankatesh24 I have tested it. The password dialog only appears for encrypted files and not for all files. Have you tested my last and latest changes?

Yeah, I had tested the latest changes only

File: app/src/main/java/swati4star/createpdf/util/PdfToImages.java [hunk 5]
```diff
@@ -26,18 +28,42 @@
...
+        }
     }
 
     @Override
     protected void onPreExecute() {
         super.onPreExecute();
         mExtractImagesListener.extractionStarted();
+        mPdfEncryptionUtility = new PDFEncryptionUtility((Activity) mContext);
+        mPDFUtils = new PDFUtils((Activity) mContext);
+        if (mUri != null)
```
Comment (vyankatesh24) [inline app/src/main/java/swati4star/createpdf/util/PdfToImages.java]: There will not be a case when `mURI` will be null. So every time `URI` is sent to that function and doing only `uri.toString()` we do not get the correct path. So I think the problem will be here.

**Comment by jddeep** [comment]
> > @vyankatesh24 I have tested it. The password dialog only appears for encrypted files and not for all files. Have you tested my last and latest changes?
> 
> Yeah, I had tested the latest changes only

@vyankatesh24 First try with an unencrypted file. (First test) Then try with encrypted files and see. Please tell me after that.

File: app/src/main/java/swati4star/createpdf/util/PdfToImages.java (hunk 5 above)
Comment (jddeep) [reply app/src/main/java/swati4star/createpdf/util/PdfToImages.java]: @vyankatesh24 No the mUri is null in most cases but the mPath is never null. Though as you have said I am removing the mUri null case. Please check.

Comment (jddeep) [reply app/src/main/java/swati4star/createpdf/util/PdfToImages.java]: I have logged then and have seen.

Comment (vyankatesh24) [reply app/src/main/java/swati4star/createpdf/util/PdfToImages.java]: Have you tested the changes?
Is it working?

Comment (jddeep) [reply app/src/main/java/swati4star/createpdf/util/PdfToImages.java]: Even after removing it. It has the same kind of problem. Though the password dialog doesn't appear  for non-encrypted files which was the case before also as i said earlier.

File: app/src/main/java/swati4star/createpdf/util/PDFEncryptionUtility.java [hunk 6]
```diff
@@ -24,6 +24,7 @@
 import swati4star.createpdf.R;
 import swati4star.createpdf.database.DatabaseHelper;
 import swati4star.createpdf.interfaces.DataSetChanged;
+import swati4star.createpdf.interfaces.OnPDFDecryptInterface;
```
Comment (vyankatesh24) [inline app/src/main/java/swati4star/createpdf/util/PDFEncryptionUtility.java]: @jddeep There are compilation errors.

**Comment by jddeep** [comment]
@vyankatesh24 Fixed!

**Comment by vyankatesh24** [comment]
> @vyankatesh24 Fixed!

Still There are in `PdfToImages.java`

**Comment by jddeep** [comment]
@vyankatesh24 Where?

**Comment by jddeep** [comment]
@vyankatesh24 Everything is fixed. Please review.

File: app/src/main/java/swati4star/createpdf/util/PdfToImages.java [hunk 7]
```diff
@@ -26,35 +28,61 @@
...
     @Override
     protected void onPreExecute() {
         super.onPreExecute();
+        mPDFEncryptionUtility = new PDFEncryptionUtility((Activity) mContext);
         mExtractImagesListener.extractionStarted();
     }
 
     @Override
     protected Void doInBackground(Void... voids) {
+        if (mPassword != null) {
```
Comment (vyankatesh24) [inline app/src/main/java/swati4star/createpdf/util/PdfToImages.java]: If you are checking here `mPassword != null` then there is really a need of another constructor and checking at line `229` of `PdfToImageFragment` ?

File: app/src/main/java/swati4star/createpdf/util/PDFEncryptionUtility.java [hunk 8]
```diff
@@ -191,6 +192,42 @@ public void afterTextChanged(Editable input) {
...
+            finalOutputFile = mFileUtils.getUniqueFileName
+                    (file.replace(mContext.getResources().getString(R.string.pdf_ext),
+                            mContext.getString(R.string.decrypted_file)));
+            password = reader.computeUserPassword();
+            byte[] input = inputPassword[0].getBytes();
+            if (Arrays.equals(input, password)) {
+                PdfStamper stamper = new PdfStamper(reader, new FileOutputStream(finalOutputFile));
+                stamper.close();
+                reader.close();
+                if (dataSetChanged != null)
```
Comment (vyankatesh24) [inline app/src/main/java/swati4star/createpdf/util/PDFEncryptionUtility.java]: What is the use of this? I did not get.

File: app/src/main/java/swati4star/createpdf/util/PDFEncryptionUtility.java [hunk 9]
```diff
@@ -191,6 +192,42 @@ public void afterTextChanged(Editable input) {
...
+                    dataSetChanged.updateDataset();
+                new DatabaseHelper(mContext).insertRecord(finalOutputFile, mContext.getString(R.string.decrypted));
```
Comment (vyankatesh24) [inline app/src/main/java/swati4star/createpdf/util/PDFEncryptionUtility.java]: Remove this line and also delete a decrpyted file in `postexecute`

File: app/src/main/java/swati4star/createpdf/fragment/PdfToImageFragment.java [hunk 10]
```diff
@@ -177,13 +183,57 @@ public void onActivityResult(int requestCode, int resultCode, Intent data) throw
...
     @OnClick(R.id.createImages)
     public void parse() {
-        if (mOperation.equals(PDF_TO_IMAGES))
-            new PdfToImages(mPath, mUri, this).execute();
-        else
+        if (mPDFUtils.isPDFEncrypted(mPath)) {
+            mInputPassword = new String[1];
+            new MaterialDialog.Builder(mActivity)
+                    .title(R.string.enter_password)
+                    .content(R.string.decrypt_message)
```
Comment (vyankatesh24) [inline app/src/main/java/swati4star/createpdf/fragment/PdfToImageFragment.java]: Change content text to `Selected pdf is password protected. Enter the password to continue...`

File: app/src/main/java/swati4star/createpdf/fragment/PdfToImageFragment.java [hunk 11]
```diff
@@ -177,13 +183,57 @@ public void onActivityResult(int requestCode, int resultCode, Intent data) throw
...
+                    .input(null, null, (dialog, input) -> {
```
Comment (vyankatesh24) [inline app/src/main/java/swati4star/createpdf/fragment/PdfToImageFragment.java]: Make it password input, not the plain text

File: app/src/main/java/swati4star/createpdf/fragment/PdfToImageFragment.java [hunk 12]
```diff
@@ -177,13 +183,57 @@ public void onActivityResult(int requestCode, int resultCode, Intent data) throw
...
+                        if (StringUtils.isEmpty(input)) {
+                            showSnackbar(mActivity, R.string.snackbar_name_not_blank);
+                        } else {
+                            final String inputName = input.toString();
+                            if (inputName != null) {
+                                mInputPassword[0] = inputName;
+                                pdfToImage(mInputPassword);
+                            } else {
+                                MaterialDialog.Builder builder = createOverwriteDialog(mActivity);
```
Comment (vyankatesh24) [inline app/src/main/java/swati4star/createpdf/fragment/PdfToImageFragment.java]: When this dialog shows up?

File: app/src/main/java/swati4star/createpdf/fragment/PdfToImageFragment.java [hunk 13]
```diff
@@ -177,13 +183,57 @@ public void onActivityResult(int requestCode, int resultCode, Intent data) throw
...
+        } else {
+            pdfToImage(mInputPassword);
+        }
+    }
+
+    /**
+     * Thia method handles the call to the Async process of conversion
+     * from PDF to Image.
+     *
+     * @param mInputPassword is the password if the file is encrypted.
```
Comment (vyankatesh24) [inline app/src/main/java/swati4star/createpdf/fragment/PdfToImageFragment.java]: Put `-` after the variable name instead of `is`

**Comment by jddeep** [comment]
@vyankatesh24 Please check into the issue label. It took quite a lot of work and time.

**Comment by jddeep** [comment]
@vyankatesh24 Changes have been made. Please review.

File: app/src/main/res/values/strings.xml [hunk 14]
```diff
@@ -200,6 +200,7 @@
     <string name="set_as_default">Set as default</string>
     <string name="decrypted_file">_unencrypted.pdf</string>
     <string name="decrypt_message">Enter the password to continue</string>
+    <string name="decrypt_protected_file">Selected pdf is password protected. Enter the password to continue...</string>
```
Comment (vyankatesh24) [inline app/src/main/res/values/strings.xml]: Where you have used it?

File: app/src/main/java/swati4star/createpdf/util/PDFEncryptionUtility.java [hunk 15]
```diff
@@ -191,6 +191,39 @@ public void afterTextChanged(Editable input) {
         });
     }
 
+    /**
+     * This function removes the password for encrypted files.
+     * @param file - the path of the actual encrypted file.
+     * @param dataSetChanged - the DataSetChanged object.
```
Comment (vyankatesh24) [inline app/src/main/java/swati4star/createpdf/util/PDFEncryptionUtility.java]: What is the use of this parameter then?

File: app/src/main/java/swati4star/createpdf/util/PDFEncryptionUtility.java [hunk 16]
```diff
@@ -191,6 +191,39 @@ public void afterTextChanged(Editable input) {
...
+            finalOutputFile = mFileUtils.getUniqueFileName
+                    (file.replace(mContext.getResources().getString(R.string.pdf_ext),
+                            mContext.getString(R.string.decrypted_file)));
+            password = reader.computeUserPassword();
+            byte[] input = inputPassword[0].getBytes();
+            if (Arrays.equals(input, password)) {
+                PdfStamper stamper = new PdfStamper(reader, new FileOutputStream(finalOutputFile));
+                stamper.close();
+                reader.close();
+                String filepath = finalOutputFile;
```
Comment (vyankatesh24) [inline app/src/main/java/swati4star/createpdf/util/PDFEncryptionUtility.java]: can directly return` finalOutputFile` ?

**Comment by jddeep** [comment]
@vyankatesh24 Have a look now.

**Pull request merged**